libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -693,23 +693,35 @@ void AP_GPS::update(void)
                 // handle switch between real GPSs
                 for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
                     if (i == primary_instance) {
-                        continue;
+                        continue; 
                     }
                     if (state[i].status > state[primary_instance].status) {
                         // we have a higher status lock, or primary is set to the blended GPS, change GPS
-                        primary_instance = i;
-                        _last_instance_swap_ms = now;
-                        continue;
+                        
+                        if (state[primary_instance].status < GPS_OK_FIX_3D) {
+                            primary_instance = i;
+                            _last_instance_swap_ms = now;
+                            continue;
+                        }
+                        // switch only if the currently used GPS has a 2D FIX or lower
+
+                        if (state[i].status > GPS_OK_FIX_3D) {
+                            primary_instance = i;
+                            _last_instance_swap_ms = now;
+                            continue;
+                        }
+                        // AND only if the other GPS has a DGPS FIX (the best possible fix)
```

Comment:
Is this logic currently really an AND logic? Looks to me as if the GPS will switch if the current primary is worse than 3D Fix OR the new primary is better than 3D fix. Doesn't this mean that you could end up switch to the new GPS if both had less than 3D fix?

Reply:
No, it is not an 'AND' logic. In the comments, I wrote 'AND' just to showcase another scenario in which it will switch. I did not mean the 'AND' by the 'AND logic operator'.
So the logic now states that, if the fix type of the other GPS is better than the fix type of the primary GPS, it will look into the possibility of switching to the other GPS. And once that condition is met, there are only 2 possibilities in which the GPS switching will take place, i.e., 
1. It will switch only when the currently used GPS has a fix type of 2D or less than that (which we absolutely want to avoid).
2. If the fix type of the other GPS which is not being used that instant has an SBAS fix, which is the best fix possible and we should always be seeking for.

Reply:
I don't see any timing considerations - it looks to me that it's possible that we could end up switching back and forth rapidly between the two receivers when they're both SBAS. I think that in the situation where both receivers have an SBAS fix type, there should be no switching or switching takes place only if the quality of the non-primary receiver is consistently better than the primary instance for an extended period of time.

Reply:
@cvh10  You are missing the main piece here. It won't even go into the switching logic unless the other GPS has a better Fix then the one used currently. So if the primary GPS has an SBAS fix, which is the maximum, it won't even go into the switching logic. 
So No, it won't keep switching between the 2 GPS's if both of them have an SBAS fix.